### PR TITLE
NEW: pagination and filtering through search

### LINKED ISSUE #156: Implement pagination when fetching all the plios (full objects or just UUIDs)
Currently, if a 100 plios exist, the client fetches all of them at once while rendering the homepage

Ideally, the plios should be fetched a page at a time and the user should be able to navigate between different pages, making a separate call for each page

### LINKED ISSUE #157: Implement filtering the plios using a search query
After pagination has been implemented, the current configuration to search plios in the homepage will be redundant.

The user should enter a search string and when the call is made to the backend, the backend should filter the plios based on that search string and return the results.

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #156 
Fixes #157

Requires [this](https://github.com/avantifellows/plio-frontend/pull/236) to be merged

TODO: Test on staging

### CODE REVIEW

File: plio/views.py [hunk 1]
```diff
@@ -81,10 +122,15 @@ def perform_update(self, serializer):
...
-            return Response([])
-        return Response(list(queryset.values_list("uuid", flat=True)))
+        queryset = self.filter_queryset(self.get_queryset())
+        uuid_list = queryset.values_list("uuid", flat=True)
+        page = self.paginate_queryset(uuid_list)
+        if page is not None:
+            return self.get_paginated_response(page)
+
+        return Response(
+            {"detail": "Cannot paginate plio uuids"}, status=status.HTTP_404_NOT_FOUND
```
Comment (dalmia) [inline plio/views.py:132]:
```suggestion
            {"detail": "No plios found"}, status=status.HTTP_404_NOT_FOUND
```

File: plio/views.py [hunk 2]
```diff
@@ -81,10 +122,15 @@ def perform_update(self, serializer):
     @action(detail=False, permission_classes=[IsAuthenticated])
     def list_uuid(self, request):
         """Retrieves a list of UUIDs for all the plios"""
-        queryset = self.get_queryset()
-        if not queryset or not queryset.exists():
-            return Response([])
-        return Response(list(queryset.values_list("uuid", flat=True)))
+        queryset = self.filter_queryset(self.get_queryset())
```
Comment (dalmia) [inline plio/views.py:125]: shouldn't the check for `if not queryset or not queryset.exists()` happen before `filter_queryset`? 
Has this been tested for ids with no plios to list?

File: plio/views.py [hunk 3]
```diff
@@ -15,6 +17,33 @@
...
+    """
+    Splits result sets into individual pages of data.
+    Pagination links are provided as part of the content
+    of the response.
+
+    Reference: django-rest-framework.org/api-guide/pagination/
+    """
+
+    # number of results in a page
+    page_size = 5
```
Comment (dalmia) [inline plio/views.py:30]: should be set in some config file no?

File: plio/views.py [hunk 4]
```diff
@@ -15,6 +17,33 @@
...
+        return Response(
+            OrderedDict(
+                [
+                    ("count", self.page.paginator.count),
+                    ("page_size", self.get_page_size(self.request)),
+                    ("next", self.get_next_link()),
+                    ("previous", self.get_previous_link()),
+                    ("results", data),
+                ]
+            )
```
Comment (dalmia) [inline plio/views.py:43]: why not this?
```suggestion
            {
                    "count": self.page.paginator.count,
                    "page_size": self.get_page_size(self.request),
                    "next": self.get_next_link(),
                    "previous": self.get_previous_link(),
                    "results": data
            }
```

File: plio/views.py [hunk 5]
```diff
@@ -128,8 +125,15 @@ def list_uuid(self, request):
         if page is not None:
             return self.get_paginated_response(page)
 
+        # return an empty response in a paginated format if pagination fails
```
Comment (dalmia) [inline plio/views.py:128]:
```suggestion
        # return an empty response in the paginated format if pagination fails
```

File: plio/views.py [hunk 6]
```diff
@@ -128,8 +125,15 @@ def list_uuid(self, request):
...
         return Response(
-            {"detail": "Cannot paginate plio uuids"}, status=status.HTTP_404_NOT_FOUND
+            {
+                "count": 0,
+                "page_size": self.get_page_size(self.request),
```
Comment (dalmia) [inline plio/views.py:132]: shouldn't this be 0?

Comment (deepansh96) [reply plio/views.py:132]: Nhi page size has already been configured in the Paginator class. Writing it 0 here will be incorrect

**Review by dalmia — APPROVED** [review]

**Pull request merged**